internal/wrphandlers/qos/priority_queue.go

```diff
@@ -0,0 +1,103 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"container/heap"
+
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+// PriorityQueue implements heap.Interface and holds wrp Message, using wrp.QOSValue as its priority.
+// https://xmidt.io/docs/wrp/basics/#qos-description-qos
+type PriorityQueue struct {
+	// heap for wrp messages, ingested by serviceQOS
+	heap []wrp.Message
+	// maxHeapSize is the allowable max size of the queue based on the sum of all queued wrp message's payloads
+	maxHeapSize int
+}
+
+// Dequeue returns the next highest priority message.
+func (pq *PriorityQueue) Dequeue() (wrp.Message, bool) {
+	// Required, otherwise heap.Pop will panic during an internal Swap call.
+	if pq.Len() == 0 {
+		return wrp.Message{}, false
+	}
+
+	msg, ok := heap.Pop(pq).(wrp.Message)
+
+	return msg, ok
+}
+
+// Enqueue queues the given message.
+func (pq *PriorityQueue) Enqueue(msg wrp.Message) {
```

Comment:
I haven't yet added the zero behavior for `maxHeapSize`
https://github.com/xmidt-org/xmidt-agent/pull/104/files#r1578469699